### ISSUE TITLE
occasionally crash

### ISSUE THREAD

**Issue by xmyqsh** [traveller59/spconv#147]
Hello,

I occasionally got the following error when using spconv. It almost works well when I disconnect my internet connection.

```
[33750/39521]
cudahash: Completely failed to build
Cuda error in file '/xxxxxx/spconv/src/cuhash/hash_table.cpp' in line 194 : an illegal memory access was encountered.
```

Any advice for this error?

By the way, I'm planning to wrap this repo into a TensorFlow version.
Hope for further help here.

**Comment by traveller59** [comment]
cuhash may have problem. currently it's recommend to disable cuda hash by set ```use_hash=False```.
Tensorflow:
implement spconv in tensorflow is straightforward but not "wrap", you still need to write much code because the tensorflow c++ operator is quite different from pytorch.

**Comment by xmyqsh** [comment]
That's true.
That's to say you recommend v1.0 instead of v1.1?

**Comment by xmyqsh** [comment]
@traveller59 
Indeed, it is straightforward.

![image](https://user-images.githubusercontent.com/16010828/83594441-ca3c1e80-a591-11ea-9843-03194b10756d.png)

What are you thinking about the description above?
30 sparse 3D conv, 7 sparse pooling, 7 sparse unpooling on 200k waymo lidar points only use 12ms.
Can you summarize the key factor in the above description that speed up the speed that spconv do not have?

**Comment by traveller59** [comment]
1. I don't know how many voxels generated from 200k waymo pointcloud.
2. I don't know the kernel size/stride of pooling operation. number of points generated from 3/2/1(k,s,p) is greatly higher than 2/2/0(ksp).
3. they seems not use regular sparse convolution. this can bring speed up.
4. their pooling operations use a better algorithm to achieve better speed.
5. they seems use fused gemm to calculate conv. I'm not sure whether fused gemm can achieve better performance.

**Comment by traveller59** [comment]
The subm indice generation speed is greatly increased (4x faster) in newest code. you can have a try.

**Comment by xmyqsh** [comment]
Interesting! Follow you!

Do you think voxel generation could be speed up by CUDA? DOPS uses pooling op to do it.

For 2, I guess they do not use 2/2/0, because 2/2/0 have poor receptive field in their UNet, and the receptive field is more harder to achieve in sparse conv net compared to normal conv net.

What is fused gemm? Can you have a brief introduction?

**Comment by traveller59** [comment]
Yes, voxel generation (one point per voxel) is equivalent to max pool operation. the maxpool in spconv isn't optimized. I will try to optimize it but I'm not sure what the "shuffle" mean in their paper.

fused gemm: perform gather-gemm-scatter in one kernel, no additional buffer needed. you can check [fused gemm](https://github.com/facebookresearch/SparseConvNet/blob/master/sparseconvnet/SCN/CUDA/Convolution.cu#L59) implemented in SparseConvNet.

**Comment by xmyqsh** [comment]
It seems that I have to revert my understanding of 2. You are correct, they use 2/2/0(ksp) instead of 3/2/1(k,s,p) for pooling.
pool 3/2/1(k,s,p) is same as conv with 3/2/1(k,s,p). But with pool 2/2/0(ksp) , each voxel will be looked up only once as the DOPS said. 
The voxel generation is same as the 2/2/0(ksp) version, which is s/s/0(ksp). And the last thing that we should make voxel generation same as 2/2/0(ksp) version of pool is that the s in s/s/0(ksp) is discrete, but the points' position is continue.
There is two way to resolve this problem, and both of them use the `only look up once property`: 
Solution 1: sort/shuffle points by the radix sort: use radix sort on the first several digits of the points' position to make it sorted in terms of stride resolution. And then do max or ave on the slot/bucket of radix sort. And CUB library has the radix sort API.
Solution 2: use HashMultiMap: your cpu single thread version of voxelization is exactly same as the HashMultiMap, the only thing to do is how to use CUDA to implemented it. Maybe several Atomic operators will resolve the problem.

So, the meaning of `shuffle` is the order of sequence unware, it is another description of `only look once property`

**Comment by traveller59** [comment]
Solution 1: I check the cub radix sort but no API found for slot access. I still don't know how to use radix to perform pooling.
Solution 2: looks ok.

I create a [benchmark script](https://github.com/traveller59/spconv/blob/develop/test/benchmark.py) with a encoder network and waymo data, it costs 40ms in 1080Ti (1ms subm indice gen, 35ms subm and 4ms maxpool). I really don't think they can achieve 12ms with both encoder and decoder... The voxel size and point cloud range all missing in their paper.

**Comment by xmyqsh** [comment]
1ms for 7 subm indice gen?
In my intuition, indice gen will consume lots of time so we need indice_key.
This is a good and helpful benchmark, thank you!
40 / 4 = 10ms
40 / 8 = 5ms
Amazing result!!
I have to say, spconv has the change to be faster than DOPS.

But you haven't use shared memory and bit operation to partitioning and caching the filter.

Have you tested your normal sparse conv compared with subm?

**Comment by traveller59** [comment]
What does "40 / 4 = 10ms" mean?

The "shared memory and bit operation to partitioning and caching the filter" is used for gemm. the fused gemm implementation in SparseConvNet use shared memory to cache filter, but spconv use cuBLAS gemm, so we don't need cache for filter.

speed of normal sparse conv isn't comparable with subm because normal conv generate too much new points and will greatly impact speed of following layers.

I will bench SparseConvNet fused gemm in this benchmark first, then consider to modify CUTLASS to get a better fused gemm.

**Comment by xmyqsh** [comment]
40 / 4 > 10ms
When we downsample the resolution, the sparsity will increase.

You are a big god, I will learn lots of things here!

**Comment by xmyqsh** [comment]
@traveller59 
Can I get the following conclusion?
Because of your efficient indice gen implementation, for a non-sparse point cloud, aka, sparsity=0, one normal/subm sparse conv has similar performance to the normal/dense conv, because both of them use GEMM internal.

Two double about the conclusion above:
1. normal/dense conv may use cudnn for acceleration.
2. with the sparsity decrease to zero, the speed of indice gen will also decrease. What is the indice  gen speed in terms of the sparsity. How about sparsity approaching zero.

Can you have an estimate that what is the sparsity when the sparse conv has the same (slow)speed as the normal/dense conv?

**Comment by xmyqsh** [comment]
Another question:
```
self.net = spconv.SparseSequential(
      spconv.SparseConv3d(32, 64, 3, 1, indice_key="cp0"),
      spconv.SparseConv3d(32, 64, 3, 1, indice_key="cp0"),
      spconv.SparseConv3d(32, 64, 3, 1, indice_key="cp0"),
)
```
This is a wrong demo of using normal `SparseConv3d` with `indice_key`, because `SparseConv3d` is different from `SubM`, the indice should be enlarged between `SparseConv3d`.

Please help me confirm this conclusion.

**Comment by xmyqsh** [comment]
How is going on?
```
#define RULEBOOKITERATOR(X, Y)                                                 \
  {                                                                            \
    Int rbMaxSize = 0;                                                         \
    for (auto &r : _rules)                                                     \
      rbMaxSize = std::max(rbMaxSize, (Int)r.size());                          \
    at::Tensor rulesBuffer = at::empty({rbMaxSize}, at::CUDA(at_kINT));        \
    Int *rbB = rulesBuffer.data_ptr<Int>();                                        \
    for (int k = 0; k < _rules.size(); ++k) {                                  \
      auto &r = _rules[k];                                                     \
      Int nHotB = r.size() / 2;                                                \
      if (nHotB) {                                                             \
        cudaMemcpy(rbB, &r[0], sizeof(Int) * 2 * nHotB,                        \
                   cudaMemcpyHostToDevice);                                    \
        X                                                                      \
      }                                                                        \
      Y                                                                        \
    }                                                                          \
  }
```

```
      if (nHotB) {                                                             \
        cudaMemcpy(rbB, &r[0], sizeof(Int) * 2 * nHotB,                        \
                   cudaMemcpyHostToDevice);                                    \
        X                                                                      \
      }
```

Do you think the copy of rulebook from host to device is the bottleneck of SparseConvNet?
A simple opt could copy it totally once instead of kernelVolumes times.

But one thing maybe concluded that comparing with speed, SparseConvNet care GPU memory usage more.

Anyway, all of data in spconv is in device, we do not have the problem.

Kernel calls comparison:
spconv(gather-mm_out-scatter): kernelVolume * 3
vs
SparseConvNet(convA(main part of nHot)---convB(remainder)): kernelVolume * 2


memory copy comparison:
spconv(gather copy + scatter implicit copy): all copy are device from device, but buffer is too large
vs  
SparseConvNet(rulebook copy (host to device) + shared memory copy(O(K^2))): shared memory is fast, but rulebook copy from host to device is too slow

All of the copy above are kernelVolume times.

memory usage comparison: 
spconv's gather/scatter's buffer is too large and device to device copy
SparseConvNet uses shared memory to do this should be faster.

Conclusion:
spconv could try to borrow SparseConvNet's shared memory conv
Benefit: shared memory copy is faster, memory usage is smaller without gather/scatter large buffer.
cons: speed is doublable which should be benchmark further.
futher: bit operation for further opt

**Comment by traveller59** [comment]
The major bottleneck of SparseConvNet is their fused gemm. the benchmark I posted before costs 40ms in spconv, but 300ms when using fused gemm in SparseConvNet instead of gather-cuBLAS-scatterAdd in spconv...

**Comment by xmyqsh** [comment]
`Grids out based GPU indice generation` surpass `hashmap based CPU indice generation`.
_________________________
Tiled matrix product makes extensive use of the register file to hold fragments and accumulator tiles as well as large shared memory allocations. Relatively high demand of on-chip storage limits occupancy, the maximum number of thread blocks that may run concurrently on one SM. 
Consequently, GEMM implementations can fit far fewer warps and thread blocks in each SM than is typical for GPU computing workloads. We use software pipelining to **hide data movement latency by executing all stages of the GEMM hierarchy concurrently within a loop and feeding the output of each stage to its dependent stage during the next iteration** as Figure 8 shows.

CUTLASS looks good, can we remove gather/scatter by using it?

**Comment by xmyqsh** [comment]
For purely matrix multiplication(DGEMM in conv), CUTLASS only faster than cuBLAS by around 5%.
Hope the flexibility Template support in CUTLASS can remove the gather/scatter op.

**Comment by xmyqsh** [comment]
Have tested the latency of gather/scatter block?
I've noticed that the gather/scatter kernel uses a similar CUDA kernel configuration as fused gemm does in SparseConvNet.

**Comment by xmyqsh** [comment]
Got it!
Let's separate the work step by step:
First, write a [DGEMM](https://www.fz-juelich.de/SharedDocs/Downloads/IAS/JSC/EN/slides/cuda/10-cuda-dgemm-tiled.pdf?__blob=publicationFile) based on `tiled matrix multiplication with cuda` algo first.
Then, benchmark its performance.
If the performance comparable with cuBLAS, develop fused gemm based on it;
else, turn to CUTLASS.

This is what SparseConvNet does exactly. Their fused gemm is modified from `tiled matrix multiplication`.
That's why I insist on benchmark it by replacing its cpu hashmap rulebook generation algo by spconv gridouts based GPU indice pair generation.
